### PR TITLE
Updates on ASR with diarization util files

### PR DESCRIPTION
**Pull request by tango4j**
This PR includes the following changes:
- Added CTM file based (word-by-word) diarization evaluation for calculating accurate WDERs.
- Created decoder_timestamps_utils.py that contains all the word time-stamp generation scripts.
- pyctcdecode support in decoder_timestamps_utils.py.
- LM based diarization transcript realigning in diarization_utils.py.
- offline_diarization_with_asr.py example and its yaml file.
- Updates on diarization README.md.
- Fixed unique ID indexing bug in scripts/speaker_tasks/pathsfiles_to_manifest.py.

### CODE REVIEW

File: examples/speaker_tasks/diarization/README.md [hunk 1]
```diff
@@ -73,25 +73,48 @@ Some of important options in config file:
...
+Single-scale setting:
+```bash
+  python offline_diarization.py \
+     ... <other paramerters> ...
+     parameters.window_length_in_sec=1.5 \
+     parameters.shift_length_in_sec=0.75 \
+     parameters.multiscale_weights=null \
+```
+
+Multiscale setting (base scale - window_length 0.5 s and shift_length 0.25):
```
Comment (fayejf) [inline examples/speaker_tasks/diarization/README.md:109]: just wondering is it necessary to introduce base scale concept?

File: examples/speaker_tasks/diarization/README.md [hunk 2]
```diff
@@ -186,3 +208,43 @@ Example: `./demo_asr_output/pred_rttms/my_audio1.txt`
 [00:13.97 - 00:15.78] speaker_1: but it's the fourth of july mm
 [00:16.90 - 00:21.80] speaker_0: so yeahg people still work tomorrow do you have to work tomorrow did you drive off yesterday
 ```
+ 
+### Optional Features for Speech Recognition with Speaker Diarization
+ 
+#### Beam Search Decoder
```
Comment (fayejf) [inline examples/speaker_tasks/diarization/README.md:214]: Is it possible to move beam search decoder related stuff to ASR for wider usage?

File: nemo/collections/asr/parts/utils/decoder_timestamps_utils.py [hunk 3]
```diff
@@ -0,0 +1,761 @@
...
+    ):
+
+        super().__init__(tokenizer, batch_dim_index, use_cer, ctc_decode, log_prediction, dist_sync_on_step)
+
+    def ctc_decoder_predictions_tensor_with_ts(
+        self, time_stride, predictions: torch.Tensor, predictions_len: torch.Tensor = None
+    ) -> List[str]:
+        hypotheses, timestamps, word_timestamps = [], [], []
+        # '⁇' string should removed since it causes error on string split.
+        unk = '⁇'
```
Comment (fayejf) [inline nemo/collections/asr/parts/utils/decoder_timestamps_utils.py:74]: why the hypothesis contains '??' what it represents?

File: nemo/collections/asr/parts/utils/diarization_utils.py [hunk 4]
```diff
@@ -748,84 +426,288 @@ def write_json_and_transcript(
...
+                log_p = [hyp_former, hyp_latter]
+                p_order = np.argsort(log_p)[::-1]
+                if log_p[p_order[0]] > log_p[p_order[1]] + self.realigning_lm_params['logprob_diff_threshold']:
+                    if p_order[0] == 0:
+                        spk_list[k] = org_spk_list[k + 1]
+                line_dict['speaker_label'] = spk_list[k]
+            realigned_list.append(line_dict)
+        return realigned_list
+
+    def get_alignment_errors(self, ctm_content, hyp_w_dict_list, mapping_dict):
```
Comment (fayejf) [inline nemo/collections/asr/parts/utils/diarization_utils.py:626]: Maybe we can move and gather evaluation codes, i.e. alignement error, der,  in another utils files to slim diarization utils later

File: scripts/speaker_tasks/pathsfiles_to_manifest.py [hunk 5]
```diff
@@ -123,7 +147,7 @@ def main(wav_scp, text_scp=None, rttm_scp=None, uem_scp=None, ctm_scp=None, mani
     parser.add_argument("--paths2rttm_files", help="path to text file containing list of rttm files", type=str)
     parser.add_argument("--paths2uem_files", help="path to uem files", type=str)
     parser.add_argument("--paths2ctm_files", help="path to ctm files", type=str)
-    parser.add_argument("--manifest_filepath", help="scp file name", type=str, required=True)
+    parser.add_argument("--manifest_filepath", help="output manifest file name", type=str, required=True)
```
Comment (fayejf) [inline scripts/speaker_tasks/pathsfiles_to_manifest.py:150]: could you change files name to file path to reduce confusion

**Review by fayejf — CHANGES_REQUESTED** [review]
Initial review: Overall great job!  Thanks! 
Some comments and could you include new tests that cover this PR's change and update previous tests if necessary?

File: nemo/collections/asr/parts/utils/decoder_timestamps_utils.py [hunk 6]
```diff
@@ -0,0 +1,761 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import math
+import sys
+from datetime import datetime
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/decoder_timestamps_utils.py:18]: not used please remove

File: nemo/collections/asr/parts/utils/decoder_timestamps_utils.py [hunk 7]
```diff
@@ -0,0 +1,761 @@
...
+from nemo.collections.asr.parts.utils.streaming_utils import AudioFeatureIterator, FrameBatchASR
+from nemo.collections.common.tokenizers.tokenizer_spec import TokenizerSpec
+from nemo.utils import logging
+
+__all__ = ['ASR_TIMESTAMPS']
+
+try:
+    from pyctcdecode import build_ctcdecoder
+except ImportError:
+    logging.info("pyctcdecode is not installed. You must install kenlm and pyctcdecode to use language models")
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/decoder_timestamps_utils.py:41]: Check for pyctcdecode only when LM scoring being requested by user (Line: 373) if not found trough error otherwise with comments to install from certain location manually.

File: examples/speaker_tasks/diarization/offline_diarization_with_asr.py [hunk 8]
```diff
@@ -47,35 +47,25 @@ def main(cfg):
 
     logging.info(f'Hydra config: {OmegaConf.to_yaml(cfg)}')
 
-    asr_diar_offline = ASR_DIAR_OFFLINE(**cfg.diarizer.asr.parameters)
-    asr_diar_offline.root_path = cfg.diarizer.out_dir
-    shutil.rmtree(asr_diar_offline.root_path, ignore_errors=True)
+    # ASR inference for words and word timestamps
+    asr_ts_decoder = ASR_TIMESTAMPS(**cfg.diarizer)
+    asr_model = asr_ts_decoder.set_asr_model(cfg.diarizer.asr.model_path)
```
Comment (nithinraok) [inline examples/speaker_tasks/diarization/offline_diarization_with_asr.py:52]: minor: ASR_model_name is already being assigned during the initialization of ASR_TIMESTAMPS. You may call this without arg and use self.ASR_model_name directly in function.

File: nemo/collections/asr/parts/utils/diarization_utils.py [hunk 9]
```diff
@@ -15,500 +15,194 @@
...
-from nemo.collections.asr.parts.utils.streaming_utils import AudioFeatureIterator, FrameBatchASR
-from nemo.collections.common.tokenizers.tokenizer_spec import TokenizerSpec
 from nemo.utils import logging
+from nemo.utils.decorators.experimental import experimental
 
-__all__ = ['ASR_DIAR_OFFLINE']
+try:
+    import arpa
+except ImportError:
+    logging.info("arpa is not installed. You must install arpa to refine diarization result with LM.")
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/diarization_utils.py:42]: You are checking for this twice.

File: nemo/collections/asr/parts/utils/diarization_utils.py [hunk 10]
```diff
@@ -15,500 +15,194 @@
...
-            word_ts[p] = [
-                max(round(word_ts[p][0] - self.asr_delay_sec, 2), 0),
-                round(word_ts[p][1] - self.asr_delay_sec, 2),
-            ]
-        return word_ts
+        self.stt_end_tokens = ['</s>', '<s>']
+        if 'arpa' in sys.modules:
+            return arpa.loadf(self.realigning_lm_params['arpa_language_model'])[0]
+        else:
+            logging.info("You must install arpa to refine diarization result with LM. Skipping realigning.")
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/diarization_utils.py:173]: Throw error here if user requested for realigning with LM but arpa or KenLM not present. with comments something like 
`you requested to use KenLM LM but its not installed. Please install from <LINK> and try again or run without realining LM case`

File: nemo/collections/asr/parts/utils/diarization_utils.py [hunk 11]
```diff
@@ -15,500 +15,194 @@
...
-            raise ValueError(f"ASR model name not found: {asr_model}")
-        self.params['time_stride'] = self.model_stride_in_secs
-        self.asr_batch_size = 16
-        asr_model.eval()
-
+        self.align_error_list = []
+        self.AUDIO_RTTM_MAP = audio_rttm_map(self.manifest_filepath)
         self.audio_file_list = [value['audio_filepath'] for _, value in self.AUDIO_RTTM_MAP.items()]
+        if self.realigning_lm_params['arpa_language_model']:
+            self.realigning_lm = self.load_realigning_LM()
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/diarization_utils.py:146]: why to load lm here even before diarization being performed?

File: nemo/collections/asr/parts/utils/diarization_utils.py [hunk 12]
```diff
@@ -527,71 +221,52 @@ def _get_speech_labels_from_decoded_prediction(self, input_word_ts):
...
 
         if diar_model_config.diarizer.asr.parameters.asr_based_vad:
-            self.save_VAD_labels_list(words_and_timestamps, self.audio_file_list)
+            self.save_VAD_labels_list(word_timestamps)
             oracle_manifest = os.path.join(self.root_path, 'asr_vad_manifest.json')
             oracle_manifest = write_rttm2manifest(self.VAD_RTTM_MAP, oracle_manifest)
             diar_model_config.diarizer.vad.model_path = None
             diar_model_config.diarizer.vad.external_vad_manifest = oracle_manifest
 
         oracle_model = ClusteringDiarizer(cfg=diar_model_config)
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/diarization_utils.py:249]: Can we rename this to diar_model, since it's not oracle_model always? Helps for clear understanding

File: nemo/collections/asr/parts/utils/diarization_utils.py [hunk 13]
```diff
@@ -925,80 +898,147 @@ def write_result_in_csv(self, args, WDER_dict, DER_result_dict, effective_WDER):
...
         dump_json_to_file(f'{ROOT}/pred_rttms/{uniq_id}.json', riva_dict)
+
+        logging.info(f"Writing {ROOT}/pred_rttms/{uniq_id}_gecko.json")
         dump_json_to_file(f'{ROOT}/pred_rttms/{uniq_id}_gecko.json', gecko_dict)
+
+        logging.info(f"Writing {ROOT}/pred_rttms/{uniq_id}.txt")
         write_txt(f'{ROOT}/pred_rttms/{uniq_id}.txt', string_out.strip())
+
+        logging.info(f"Writing {ROOT}/pred_rttms/{uniq_id}.w.label")
         write_txt(f'{ROOT}/pred_rttms/{uniq_id}.w.label', '\n'.join(audacity_label_words))
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/diarization_utils.py:975]: Can we not write info for each uniq_id, just info ing in which directory these are being stored would be enough, else sys out is becoming total cluttered.

**Review by nithinraok — CHANGES_REQUESTED** [review]

File: nemo/collections/asr/parts/utils/diarization_utils.py (hunk 9 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/diarization_utils.py:42]: If I don't use try/except, then it will pop-out error if arpa is not installed. If you want to remove the warning sign, then it will make more sense.

**Review by nithinraok — COMMENTED** [review]
Also please update current Asr with diarization tutorial notebook with these optional features.

File: examples/speaker_tasks/diarization/README.md (hunk 1 above)
Comment (tango4j) [reply examples/speaker_tasks/diarization/README.md:109]: I think this is necessary because the actual output of the diarization result is based on the base scale. I think we should link paper in the NeMo docs.

File: nemo/collections/asr/parts/utils/decoder_timestamps_utils.py (hunk 3 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/decoder_timestamps_utils.py:74]: '⁇' string is generated from decoder and this causes error during string process. This should be removed to make decoder function work smoothly.

File: nemo/collections/asr/parts/utils/diarization_utils.py (hunk 4 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/diarization_utils.py:626]: If that is the plan, bare-diarization is better to be arranged to one place. It is a thing to be discussed.

File: scripts/speaker_tasks/pathsfiles_to_manifest.py (hunk 5 above)
Comment (tango4j) [reply scripts/speaker_tasks/pathsfiles_to_manifest.py:150]: Changed to path to output manifestfile.

File: nemo/collections/asr/parts/utils/decoder_timestamps_utils.py (hunk 6 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/decoder_timestamps_utils.py:18]: datetime is unused library so removed.

File: examples/speaker_tasks/diarization/offline_diarization_with_asr.py (hunk 8 above)
Comment (tango4j) [reply examples/speaker_tasks/diarization/offline_diarization_with_asr.py:52]: Need to discuss this.

File: nemo/collections/asr/parts/utils/decoder_timestamps_utils.py (hunk 7 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/decoder_timestamps_utils.py:41]: Need to discuss this.

File: nemo/collections/asr/parts/utils/diarization_utils.py (hunk 9 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/diarization_utils.py:42]: Need to discuss this.

File: nemo/collections/asr/parts/utils/diarization_utils.py (hunk 12 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/diarization_utils.py:249]: Changed to diar_model

File: examples/speaker_tasks/diarization/offline_diarization_with_asr.py [hunk 14]
```diff
@@ -34,11 +33,12 @@
 
 Check out whole parameters in ./conf/offline_diarization_with_asr.yaml and their meanings.
 For details, have a look at <NeMo_git_root>/tutorials/speaker_tasks/Speaker_Diarization_Inference.ipynb
+Currently, the following NGC models are supported:
 
-Currently Supported ASR models:
+    QuartzNet15x5Base-En
+    stt_en_citrinet*
+    stt_en_conformer_ctc_*
```
[the remark this answers is not in the document]
Comment (tango4j) [reply examples/speaker_tasks/diarization/offline_diarization_with_asr.py:41]: Keeping these but adding the same list to README.

File: examples/speaker_tasks/diarization/README.md (hunk 2 above)
Comment (tango4j) [reply examples/speaker_tasks/diarization/README.md:214]: Once ASR team incorporates pyctcdecode, ctcdecode (Baidu's implementation) will be replaced with pyctcdecode so it will be automatically done.

File: nemo/collections/asr/parts/utils/diarization_utils.py (hunk 10 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/diarization_utils.py:173]: Fixed as discussed.

File: nemo/collections/asr/parts/utils/diarization_utils.py (hunk 11 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/diarization_utils.py:146]: This part is moved to the actual realignment part.

File: scripts/speaker_tasks/pathsfiles_to_manifest.py (hunk 5 above)
Comment (fayejf) [reply scripts/speaker_tasks/pathsfiles_to_manifest.py]: Could we output manifest file path instead just to be accurate?

File: nemo/collections/asr/parts/utils/diarization_utils.py (hunk 13 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/diarization_utils.py]: It is reduced to one line.

**Review by nithinraok — APPROVED** [review]

**Review by fayejf — APPROVED** [review]
LGTM! Thanks! It would be better if you could also include new changes to the description such as tutorials change. And let's discuss some remaining suggestion later.
